.github/workflows/go-dependency-check.yml

```diff
@@ -0,0 +1,34 @@
+name: Go Dependency Review
+on:
+  workflow_call:
+    inputs:
+      subdirectory:
+        type: string
+        description: The subdirectory to check dependencies
+        default: .
+        required: false
+
+jobs:
+  go-dependency-review:
+    name: Go Dependency Review
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        working-directory: ${{ inputs.subdirectory }}
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-go@v5
+        with:
+          go-version-file: ${{ inputs.subdirectory }}/go.mod
+          check-latest: true
+          cache-dependency-path: |
+            ${{ inputs.subdirectory }}/go.sum
+      - name: Install go-licenses
+        run: go install github.com/google/go-licenses@v1.6.0
+      - run: go mod download
```

Comment:
@joshuagibeonm not sure if the go-license already handle transitive dependency check.
But if not, we might need to use `go mod tidy` instead.
Taken from: https://go.dev/ref/mod#go-mod-tidy

> it adds any missing module requirements necessary to build the current module’s packages and dependencies, and it removes requirements on modules that don’t provide any relevant packages. It also adds any missing entries to go.sum and removes unnecessary entries. 

Meaning it updates the go.mod also with the transitive dependency so that license on it can also be checked.

Reply:
I would avoid running `go mod tidy` in the CI as `go mod tidy` should be run on local and then commited to the repo.
If we run `go mod tidy` in the CI, then the checked `go.mod` and `go.mod` that used for build could be different.

Reply:
@joshuagibeonm yes that's the point. The difference is the one that will be caught on the steps. My point is if go mod tidy somehow skipped locally, we can still catch the misconfiguration on the CI level

Reply:
@franzramadhan Running `go mod tidy` alone won't catch the violation and this PR does not cover that case. We could create a different workflow to check whether go modules is already tidy or not though, let me know if you want one.

Something like (snippet from [go-licenses](https://github.com/google/go-licenses/blob/d01822334fba5896920a060f762ea7ecdbd086e8/.github/workflows/go.yml#L29)):
```
- name: Mod Tidy
  run: go mod tidy && git diff --exit-code -- go.mod go.sum || (echo "go modules are not tidy, run 'go mod tidy'." && exit 1)
```
---
For transitive dependency. CMIIW but since go 1.17, `go.mod` already lists all transitive dependency. Usually denoted by #indirect comment.

Reply:
If the go-license do the transitive dependency check on go.mod then we can rely on `go mod download`

Reply:
> For transitive dependency. CMIIW but since go 1.17, go.mod already lists all transitive dependency.
@joshuagibeonm did you mean go.sum? If so, yes it includes the transitive dependency. If go-license also do check check on go.sum, then it's good then

Reply:
> The handler you mentioned above is out of scope and should be put in generic CI or testing suite jobs and not here.

That is what I exactly said before.

> Imagine if some dev forget to do mod tidy locally. Then package A that he uses include package B. Package B requires package C which is not explicitly listed. But the package C license falls under violated rules.

Then we need to tell them that they need to run `go mod tidy` (which is a good idea, we don't want to have discrepancy between commited go.mod and go.mod that used for build and CI), not safeguard this action from that possibility.

> CMIIW but using the go mod download alone won't add that package C into go.mod and the license will not be evaluated

Let me check on this one first.

Reply:
By the way, go-licenses already support transitive dependency. So we don't have to run `go mod tidy`.

Reply:
> My concern here is more on if the transitive dependency checking is already handled or not.
> If the go-license can utilize cached dependency retrieved by mod download to retrieve the transitive dependency and do the check, then it's good. If not, we need to do something.

Based on my reasoning above, LGTM then